### PR TITLE
implementation of the price hook

### PR DESCRIPTION
**Pull request by ShawTim**

### CODE REVIEW

File: packages/tempus-core-services/src/services/StatisticsService.ts [hunk 1]
```diff
@@ -116,38 +116,44 @@ export class StatisticsService {
...
-
     // TODO - Refactor getRate function to accept token precision as well as a parameter
     const precision = tokenPrecision[tokenTicker];
 
-    return div18f(rate, rateDenominator, precision);
+    return from(
+      overrides ? this.stats.getRate(chainLinkAggregator, overrides) : this.stats.getRate(chainLinkAggregator),
+    ).pipe(
+      map<[BigNumber, BigNumber], Decimal>(([rate, rateDenominator]) => {
+        const rate18f = increasePrecision(rate, 18 - (precision ?? 0));
```
Comment (danieledamiani) [inline packages/tempus-core-services/src/services/StatisticsService.ts:137]: `18` is `DEFAULT_TOKEN_PRECISION`?

File: packages/tempus-client_v3/src/hooks/useTokenRates.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+interface TokenRateMap {
+  [x: string]: Decimal | null;
+}
+
+const tokenRates$: Observable<TokenRateMap> = interval(TOKEN_RATE_POLLING_INTERVAL_IN_MS).pipe(
+  startWith(0),
+  map(() => getConfigManager().getPoolList()),
+  mergeMap<TempusPool[], Observable<TokenRateMap>>((tempusPools: TempusPool[]) => {
+    const unqiueTokenSet = new Set(
```
Comment (danieledamiani) [inline packages/tempus-client_v3/src/hooks/useTokenRates.ts:17]: `unqiueTokenSet` has a typo, should be `uniqueTokenSet`

File: packages/tempus-client_v3/src/hooks/useTokenRates.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import { getConfigManager } from '../config/getConfigManager';
+
+const TOKEN_RATE_POLLING_INTERVAL_IN_MS = 30000;
+const DEBOUNCE_IN_MS = 1000;
+
+interface TokenRateMap {
+  [x: string]: Decimal | null;
+}
+
+const tokenRates$: Observable<TokenRateMap> = interval(TOKEN_RATE_POLLING_INTERVAL_IN_MS).pipe(
```
Comment (danieledamiani) [inline packages/tempus-client_v3/src/hooks/useTokenRates.ts:13]: I think there's no need to retrieve the config every time the interval fires. A `combineLatest([interval(TOKEN_RATE_POLLING_INTERVAL_IN_MS), from(getConfigManager().getPoolList())])` could make things easier

File: packages/tempus-core-services/src/services/StatisticsService.ts (hunk 1 above)
Comment (ShawTim) [reply packages/tempus-core-services/src/services/StatisticsService.ts]: fixed

File: packages/tempus-client_v3/src/hooks/useTokenRates.ts (hunk 2 above)
Comment (ShawTim) [reply packages/tempus-client_v3/src/hooks/useTokenRates.ts]: fixed

File: packages/tempus-client_v3/src/hooks/useTokenRates.ts [hunk 4]
```diff
@@ -1,34 +1,53 @@
...
+
+interface TokenInfoMap {
+  [x: string]: {
+    chain: Chain;
+    token: Ticker;
+  };
+}
 
 interface TokenRateMap {
   [x: string]: Decimal | null;
```
Comment (danieledamiani) [inline packages/tempus-client_v3/src/hooks/useTokenRates.ts:30]: shouldn't `x` be `ticker` or `token`?

File: packages/tempus-client_v3/src/hooks/useTokenRates.ts [hunk 5]
```diff
@@ -1,34 +1,53 @@
...
+import { getServices, Decimal, StatisticsService, TempusPool, Chain, Ticker } from 'tempus-core-services';
 import { getConfigManager } from '../config/getConfigManager';
+import { config$ } from './useConfig';
 
 const TOKEN_RATE_POLLING_INTERVAL_IN_MS = 30000;
-const DEBOUNCE_IN_MS = 1000;
+const DEBOUNCE_IN_MS = 500;
+
+interface TokenInfoMap {
+  [x: string]: {
```
Comment (danieledamiani) [inline packages/tempus-client_v3/src/hooks/useTokenRates.ts:23]: see my comment below

File: packages/tempus-client_v3/src/hooks/useTokenRates.spec.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+          let price = new Decimal(0);
+          switch (tokenTicker) {
+            case 'ETH':
+              price = new Decimal(chain === 'fantom' ? 3001 : 2999);
+              break;
+            case 'USDC':
+              price = new Decimal(chain === 'fantom' ? 1.001 : 0.999);
+              break;
+            default:
+              break;
```
Comment (danieledamiani) [inline packages/tempus-client_v3/src/hooks/useTokenRates.spec.ts:35]: this `break` may be not necessary

File: packages/tempus-client_v3/src/hooks/useTokenRates.ts (hunk 4 above)
Comment (ShawTim) [reply packages/tempus-client_v3/src/hooks/useTokenRates.ts:30]: `x` here is `Chain` + `Ticker`, e.g. `ethereum-ETH`

File: packages/tempus-client_v3/src/hooks/useTokenRates.spec.ts (hunk 6 above)
Comment (ShawTim) [reply packages/tempus-client_v3/src/hooks/useTokenRates.spec.ts]: removed

File: packages/tempus-client_v3/src/hooks/useTokenRates.ts (hunk 5 above)
Comment (ShawTim) [reply packages/tempus-client_v3/src/hooks/useTokenRates.ts]: update `x` to `chainTokenString` for better naming

**Review by danieledamiani — APPROVED** [review]

**Pull request merged**